teps/0131-tekton-conformance-policy.md

```diff
@@ -0,0 +1,268 @@
+* **Pipelines Primitives**: defines workloads execution pipeline
+  * Task, TaskRun, Pipeline, PipelineRun, CustomRun.
+* **Triggers Primitives**: defines events listening and handling
+  * Triggers, TriggerBindings.
+* **Results Primitives**: defines execution history of workloads defined by Pipeline Primitives.
+  * Results, Records
+
+**Tekton API Specification**: describes ***Tekton Primitives*** in detail.
+**Tekton GA API Specifications**: describes ***GA (V1 for now) Tekton Primitives*** in detail.
+
+**Tekton Pipelines API Specification**: describes ***Tekton Pipelines Primitives*** (`Task`, `TaskRun`, `Pipeline`, `PipelineRun`, `CustomRun` of all version tracks) in detail.
+**Tekton GA Pipelines API Specification**: describes ***Tekton GA Pipelines Primitives*** (`v1.Task`, `v1.TaskRun`, `v1.Pipeline`, `v1.PipelineRun`) in detail.
+
+**Tekton Conformance API Specification**: ***core Tekton GA API Specification*** each vendor MUST implement, fields existing in the Tekton API Specification but not in the Conformance API Specification are `OPTIONAL` for the Conformance Policy by default.
+**Tekton Pipelines Conformance API Specification**: ***core Tekton GA Pipelines API Specification*** each vendor MUST implement, including core fields of `v1.TaskRun` and `v1.PipelineRun`, within where `v1.Task` and `v1.Pipeline` can be incorporated by reference.
+
+**Tekton Users**: use ***Tekton Primitives*** to define CI/CD pipelines.
+**Tekton Vendors**: provide a product built upon Tekton that conforms to the ***Tekton Conformance Policy***.
+
+## Proposal
+
+Vendors can claim **Tekton `<Primitive Type>` Conformance** on the Services they provide. For example, a vendor can claim that its product conforms to "pipelines.v1.0" Conformance Policy, which is the initial version of Tekton Pipelines V1 Conformance Policy. See this section to know details about versioning.
+
+### Conformant Services
+
+Conformant Services are services that can serve **Tekton Conformance API Specifications**. We only define Conformance API Specifications for **Tekton GA Primitives**. For now, only **Tekton Pipelines Primitives** are eligible. *Worth noting that one prerequisite for conformance is that a primitive reaches GA, not every primitive that is GA is automatically part of conformance.*
+
+Vendors must be able to accept **Tekton `<Primitive Type>` Conformance API Specification**, defined in **YAML or JSON**, submitted **in any way** as long as the schema conforms to the policy[^schema_submission], then schedule workloads, produce outputs (resource conditions and results) as declared.
+
+For now, only Tekton Pipelines Conformance API Specification is defined (APIs of Triggers and Results are out of scope): a ***Conformant Pipeline Service*** must be able to accept a `v1.PipelineRun`, `v1.TaskRun`, within where `v1.Task` and `v1.Pipeline` can be incorporated by reference.
```

Comment:
> Does this mean that a conformant service does not have to be able to host the definition of tasks and pipelines?

Yes, that's exactly what this means. And yes, vendors can chose to use remote resources instead of hosted in-cluster resources.

Reply:
I don't think we should allow for a Tekton implementation that only supports remote resources to be compliant. 
Reference to hosted resources is part of the API and has been the default mechanism from the beginning. 
Most pipelines out there would not run on a Tekton-like system that only supported remote resources.

Reply:
Hi @afrittoli, thank you for the comment, IIUC (please correct me If I'm wrong!), **Pipelines** and **Tasks** are originally designed as **static reusable components** in the Tekton ecosystem. I would imagine that storing them in the local cluster is the easiest and most intuitive approach at that time. While we use **Runtime Resources** (within where static reusable components can be incorporated by reference) to define the CI/CD. 

Requiring vendors who provide CI/CD solutions to host static reusable components does not sound very intuitive -- especially given that we have remote resolution today.

A similar case is **Docker Daemon** and **Docker Registry**: **DockerDaemon** listens for **Docker API** requests and manages Docker objects such as images, containers, networks, and volumes, while **_Docker Registry_** is the place to store reusable images.

Conformant **Tekton Pipeline Service** listens for Pipeline API request and manages **Pipelines Runtime Objects** such as `PipelineRun`s and `TaskRun`s, while **_Tekton Hub / Catalog_** is the place to store reusable pipelines/tasks.

Let me know your thoughts!

Reply:
@XinruZhang thank you for your reply.

My point was that a Tekton conformant service should support a `TaskRun` like the one we have for instance in the [getting started](https://tekton.dev/docs/getting-started/tasks/):

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  name: hello-task-run
spec:
  taskRef:
    name: hello
```

How the `hello` reference is resolved to the actual `Task` specification is up to the vendor.
A vendor may decide to store or not store the `Task` definitions as part of their service, as long as the API is supported.

Reply:
@afrittoli As a provider offering a Tekton conformant API, the example you provide is problematic because there's no way for a user to actually know which `hello` task they are running without it being resolved via remote resolution.  

Versioning is going to be very important for usability in conformance.

Reply:
@R2wenD2 my concern is that until very recently we did not have remote resolution, and so the bulk of pipelines out there use references like those in my example, so I'm inclined to say that a system that cannot run the example pipeline is not Tekton conformant. 

That said, I agree that versioning is important and I definitely would like to see remote resolution references being part of the conformance API. 

Vendors could implement/configure a default resolver that is used to resolve Tekton task references like in the example above. They could consider it a simplified version of reference, similar in a way to running a container with the `ubuntu` image. While it is a best practice to use a fully qualified container image reference, the short version (that implies both the container registry as well as the version) is still supported.

Reply:
This TEP is proposing that conformant PipelineRuns/TaskRuns should allow Pipelines/Tasks by reference. I don't think it is suggesting using a particular syntax though it is important that we figure that out when we are publishing the conformance spec

Reply:
Thank @dibyom - yes we can also continue this discussion in the next PR.